src/ppx/xr_component.cpp

```diff
@@ -48,6 +50,48 @@ static XrBool32 XrDebugUtilsMessengerCallback(XrDebugUtilsMessageSeverityFlagsEX
     return XR_FALSE; /* OpenXR spec suggests always returning false. */
 }
 
+template <size_t N>
+void CharArrayStrCpy(char (&dst)[N], std::string_view src)
+{
+    PPX_ASSERT_MSG(src.length() < N, "source string too large");
+    size_t length = std::min<size_t>(N - 1, src.length());
+    memcpy(dst, src.data(), length);
+    dst[length] = '\0';
+}
+
+float Sign(float x)
+{
+    return (x < 0 ? -1 : (x > 0 ? 1 : 0));
+}
+
+glm::quat FromXr(const XrQuaternionf& q)
+{
+    return glm::quat(q.w, q.x, q.y, q.z);
+}
+
+XrVector3f ForwardVector(const XrPosef& controller)
+{
+    const glm::vec3 unitForward = glm::vec3(0, 0, -1);
+
+    const glm::quat q   = FromXr(controller.orientation);
+    const glm::vec3 res = glm::rotate(q, unitForward);
+    return XrVector3f{res.x, res.y, res.z};
+}
+
+std::optional<XrVector2f> ProjectCursor(XrPosef controller, float z)
+{
+    XrVector3f pos = controller.position;
+    XrVector3f fwd = ForwardVector(controller);
+
+    float zDistance = z - pos.z;
+    // Controller is not pointing forward (or very sideway)
+    if (Sign(zDistance) * fwd.z < 0.001) {
```

Comment:
maybe better to add the comment here to explain that it is doing the dot product here dot(normalize(vec3(0,0,z-pos.z)), fwd)?

Reply:
Done, it's now an explicit dot product. Might take a few more cpu cycle.